File: conformance/work-class/1.0.0-draft.2/python/implementation/reference_host.py
"""Serialized single-process host foundations for draft 2."""

from __future__ import annotations

import copy
import threading

from . import reservation_v2 as reservation
from .common import canonical, digest, schema_valid


class SerializedReservationHost:
    """Own one complete reservation state and serialize submitted events.

    The class supplies the reference host's process-local ordering guarantee.
    It does not provide durable storage, external observation authenticity or
    distributed coordination.
    """

    def __init__(self, state, selected_pin):
        if not reservation._replay_state(state, selected_pin):
            raise ValueError("invalid restart state")
        self._pin = selected_pin
        self._state = copy.deepcopy(state)
        self._lock = threading.RLock()

    def snapshot(self):
        with self._lock:
            return copy.deepcopy(self._state)

    def submit(self, event, host_evidence):
        with self._lock:
            request = {
                "schema": reservation.PREFIX + "reservation-input",
                "specification_pin": self._pin,
                "state": copy.deepcopy(self._state),
                "state_digest": reservation._state_digest(self._state),
                "event": copy.deepcopy(event),
                "host_evidence": copy.deepcopy(host_evidence),
            }
            result = reservation.aggregate_step(request, self._pin)
            if result.get("status") == "TRANSITION":
                self._state = copy.deepcopy(result["state"])
            return copy.deepcopy(result)

    def restart(self, state):
        with self._lock:
            if not reservation._replay_state(state, self._pin):
                raise ValueError("invalid restart state")
            self._state = copy.deepcopy(state)


def _budget_rows(event):
    return event.get("budget_inputs", []) if isinstance(event, dict) else []


def _budget_digest_rows(rows, state_key="state"):
    result = []
    for row in rows:
        state = row[state_key]
        result.append({
            "registry_digest": row["registry_digest"],
            "affected_anchors": copy.deepcopy(row["affected_anchors"]),
            "state_digest": reservation._state_digest(state),
        })
    return sorted(result, key=lambda row: row["registry_digest"].encode("utf-8"))


class SingleProcessReferenceHost:
    """Serialize deployment, evaluation, and atomic in-process commits.

    Lifecycle policy stays in the supplied consumer. This class owns the
    authoritative roots and the private canonical replay preimages required by
    the host boundary.
    """

    def __init__(self, consumer):
        self._consumer = consumer
        self._roots = {}
        self._budgets = {}
        self._lock = threading.RLock()

    @staticmethod
    def _key(deploy_request):
        authorization = deploy_request["deployment_authorization"]
        return authorization["organization_id"], deploy_request["instance_id"]

    @staticmethod
    def _admit_budget_roots(rows, selected_pin):
        rows = list(rows)
        if len(rows) > 1:
            raise ValueError("at most one reservation registry is supported")
        admitted = {}
        for row in rows:
            try:
                registry_digest = row["registry_digest"]
                state = row["state"]
                state_digest = reservation._state_digest(state)
                anchors = row["affected_anchors"]
                invalid = (
                    registry_digest in admitted
                    or anchors != sorted(anchors, key=lambda value: value.encode("utf-8"))
                    or len(anchors) != len(set(anchors))
                    or state_digest != row["state_digest"]
                    or state["specification_pin"] != selected_pin
                    or registry_digest != digest("reservation-registry",state["core"]["configuration"])
                    or not reservation._replay_state(state, selected_pin)
                )
            except (KeyError,TypeError,ValueError):
                invalid = True
            if invalid:
                raise ValueError("invalid initial reservation root")
            admitted[registry_digest] = {
                "state": canonical(state),
                "state_value": copy.deepcopy(state),
                "state_digest": state_digest,
                "affected_anchors": copy.deepcopy(row["affected_anchors"]),
            }
        return admitted

    def install(self, deploy_request, budget_roots=()):
        """Calculate and install one authoritative initial root."""
        calculated = self._consumer.deploy(copy.deepcopy(deploy_request))
        if calculated.get("status") != "DEPLOYED":
            return copy.deepcopy(calculated)
        proposed_budgets = self._admit_budget_roots(budget_roots, calculated["state"].get("specification_pin", reservation.PIN))
        installation_budget_roots = canonical([
            {
                "registry_digest": registry_digest,
                "affected_anchors": copy.deepcopy(row["affected_anchors"]),
                "state": copy.deepcopy(row["state_value"]),
                "state_digest": row["state_digest"],
            }
            for registry_digest, row in sorted(proposed_budgets.items(), key=lambda item: item[0].encode("utf-8"))
        ])
        key = self._key(deploy_request)
        with self._lock:
            current = self._roots.get(key)
            request_bytes = canonical(deploy_request)
            state_bytes = canonical(calculated["state"])
            registry_digests = sorted(proposed_budgets, key=lambda value: value.encode("utf-8"))
            if current is not None:
                replay = (
                    current["deploy_request"] == request_bytes
                    and current["state"] == state_bytes
                    and current["installation_budget_roots"] == installation_budget_roots
                )
                if replay:
                    return {
                        "status": "HOST_DEPLOY",
                        "disposition": "INSTALLED",
                        "organization": key[0],
                        "instance_id": key[1],
                        "state_digest": current["state_digest"],
                        "replay": True,
                        "dispatch_performed": False,
                    }
                return {
                    "status": "HOST_DEPLOY",
                    "disposition": "INSTANCE_OCCUPIED",
                    "organization": key[0],
                    "instance_id": key[1],
                    "authoritative_state_digest": current["state_digest"],
                    "proposed_state_digest": calculated["state_digest"],
                    "replay": False,
                    "dispatch_performed": False,
                }
            for registry_digest, row in proposed_budgets.items():
                stored = self._budgets.get((key[0], registry_digest))
                if stored is not None and (stored["state_digest"] != row["state_digest"] or stored["state"] != row["state"]):
                    raise ValueError("reservation root revision conflict")
            for registry_digest, row in proposed_budgets.items():
                self._budgets[(key[0], registry_digest)] = row
            self._roots[key] = {
                "deploy_request": request_bytes,
                "state": state_bytes,
                "state_value": copy.deepcopy(calculated["state"]),
                "state_digest": calculated["state_digest"],
                "registry_digests": registry_digests,
                "installation_budget_roots": installation_budget_roots,
                "event_log": {},
            }
            result = {
                "status": "HOST_DEPLOY",
                "disposition": "INSTALLED",
                "organization": key[0],
                "instance_id": key[1],
                "state_digest": calculated["state_digest"],
                "replay": False,
                "dispatch_performed": False,
            }
            if not schema_valid(result, "lifecycle.schema.json", "hostDeploymentResult"):
                raise RuntimeError("host produced an invalid deployment result")
            return result

    def snapshot(self, organization, instance_id):
        with self._lock:
            root = self._roots[(organization, instance_id)]
            return {
                "state": copy.deepcopy(root["state_value"]),
                "state_digest": root["state_digest"],
                "budget_states": [
                    {
                        "registry_digest": registry_digest,
                        "affected_anchors": copy.deepcopy(row["affected_anchors"]),
                        "state": copy.deepcopy(row["state_value"]),
                        "state_digest": row["state_digest"],
                    }
                    for registry_digest in root["registry_digests"]
                    for row in (self._budgets[(organization, registry_digest)],)
                ],
            }

    def _stateful_refusal(self, organization, root, request, code):
        result = {
            "status": "REFUSED",
            "profile": request["profile"],
            "role": request["role"],
            "code": code,
            "path": "",
            "state": copy.deepcopy(root["state_value"]),
            "state_digest": root["state_digest"],
            "budget_states": [
                {
                    "registry_digest": registry_digest,
                    "affected_anchors": copy.deepcopy(row["affected_anchors"]),
                    "state": copy.deepcopy(row["state_value"]),
                    "state_digest": row["state_digest"],
                }
                for registry_digest in root["registry_digests"]
                for row in (self._budgets[(organization, registry_digest)],)
            ],
        }
        if not schema_valid(result, "lifecycle.schema.json", "statefulRefusal"):
            raise RuntimeError("host produced an invalid state refusal")
        return result

    def submit(self, organization, request, failure_reference=None):
        """Evaluate and atomically commit one event against the owned root."""
        key = (organization, request["state"]["instance_id"])
        with self._lock:
            root = self._roots.get(key)
            if root is None:
                raise ValueError("instance is not installed")
            event = request["event"]
            supplied_budgets = _budget_rows(event)
            retained = root["event_log"].get(event["event_id"])
            if retained is not None and canonical(event) != retained["event"]:
                return self._stateful_refusal(organization, root, request, "REPLAY_CONFLICT")
            if request["state_digest"] != root["state_digest"] or canonical(request["state"]) != root["state"]:
                return self._stateful_refusal(organization, root, request, "REVISION_CONFLICT")
            if retained is not None:
                result = self._consumer.step(copy.deepcopy(request))
                if result.get("status") != "STEP":
                    return copy.deepcopy(result)
                if not result["replay"]:
                    raise RuntimeError("consumer did not recognize retained exact replay")
                return {"transition": copy.deepcopy(result), "host_commit": None}
            for row in supplied_budgets:
                current = self._budgets.get((organization, row["registry_digest"]))
                supplied = row["request"]["state"]
                supplied_digest = row["request"]["state_digest"]
                if current is None or (
                    supplied_digest != current["state_digest"] or canonical(supplied) != current["state"]
                ):
                    return self._stateful_refusal(organization, root, request, "REVISION_CONFLICT")

            result = self._consumer.step(copy.deepcopy(request))
            if result.get("status") != "STEP":
                return copy.deepcopy(result)
            if result["replay"]:
                raise RuntimeError("consumer replay was not retained by the host")

            before_budgets = _budget_digest_rows([
                {
                    "registry_digest": row["registry_digest"],
                    "affected_anchors": row["affected_anchors"],
                    "state": row["request"]["state"],
                }
                for row in supplied_budgets
            ])
            after_budgets = _budget_digest_rows([
                {
                    "registry_digest": row["registry_digest"],
                    "affected_anchors": row["affected_anchors"],
                    "state": row["result"]["state"],
                }
                for row in result["budget_results"]
            ])
            if failure_reference is not None:
                failed = {
                    "status": "HOST_COMMIT",
                    "disposition": "COMMIT_FAILED",
                    "transaction_digest": result["transaction_digest"],
                    "state_before_digest": request["state_digest"],
                    "authoritative_state_digest": root["state_digest"],
                    "budget_before": before_budgets,
                    "authoritative_budgets": before_budgets,
                    "failure_reference": failure_reference,
                    "dispatch_performed": False,
                }
                if not schema_valid(failed, "lifecycle.schema.json", "hostCommitResult"):
                    raise RuntimeError("host produced an invalid commit failure")
                return {"transition": copy.deepcopy(result), "host_commit": failed}

            for row in result["budget_results"]:
                budget_state = row["result"]["state"]
                self._budgets[(organization, row["registry_digest"])] = {
                    "state": canonical(budget_state),
                    "state_value": copy.deepcopy(budget_state),
                    "state_digest": reservation._state_digest(budget_state),
                    "affected_anchors": copy.deepcopy(row["affected_anchors"]),
                }
                if row["registry_digest"] not in root["registry_digests"]:
                    root["registry_digests"].append(row["registry_digest"])
            root["registry_digests"].sort(key=lambda value: value.encode("utf-8"))
            next_log = copy.deepcopy(root["event_log"])
            next_log[event["event_id"]] = {
                "event_digest": digest("runtime-event", event),
                "transaction_digest": result["transaction_digest"],
                "event": canonical(event),
                "request": canonical(request),
                "response": canonical(result),
            }
            new_root = {
                **root,
                "state": canonical(result["state"]),
                "state_value": copy.deepcopy(result["state"]),
                "state_digest": result["state_digest"],
                "event_log": next_log,
            }
            self._roots[key] = new_root
            committed = {
                "status": "HOST_COMMIT",
                "disposition": "COMMITTED",
                "transaction_digest": result["transaction_digest"],
                "state_before_digest": request["state_digest"],
                "state_after_digest": result["state_digest"],
                "budget_before": before_budgets,
                "budget_after": after_budgets,
                "dispatch_performed": False,
            }
            if not schema_valid(committed, "lifecycle.schema.json", "hostCommitResult"):
                raise RuntimeError("host produced an invalid commit result")
            return {"transition": copy.deepcopy(result), "host_commit": committed}

File: conformance/work-class/1.0.0-draft.2/typescript/reference-host.ts
import { AdmissionError, byteCompare, canonical, digest } from './canonical.ts';
import { deploy, stateDigest, step } from './lifecycle.ts';
import { stateAdmission as reservationStateAdmission } from './reservations.ts';
import { requireShape } from './schema.ts';
import { instant } from './time.ts';
import { same } from './values.ts';
import type { RecordValue as Row } from './values.ts';

const clone = <T>(value: T): T => structuredClone(value);

export type BudgetRoot = {
  registry_digest: string;
  affected_anchors: string[];
  state: Row;
  state_digest: string;
};

type StoredBudget = BudgetRoot & { state_bytes: string };
type EventLogEntry = { event_digest: string; transaction_digest: string; event: string; request: string; response: string };
type StoredRoot = {
  deploy_request: string;
  state: Row;
  state_bytes: string;
  state_digest: string;
  registry_digests: string[];
  installation_budget_roots: string;
  event_log: Map<string, EventLogEntry>;
};

function budgetKey(organization: string, registryDigest: string): string { return canonical([organization, registryDigest]); }
function rootKey(organization: string, instanceId: string): string { return canonical([organization, instanceId]); }

export class SingleProcessReferenceHost {
  private readonly selectedPin: string;
  private readonly roots = new Map<string, StoredRoot>();
  private readonly budgets = new Map<string, StoredBudget>();
  private readonly dispatchClaims = new Set<string>();
  private queue: Promise<void> = Promise.resolve();

  constructor(selectedPin: string) { this.selectedPin = selectedPin; }

  private enqueue<T>(operation: () => Promise<T> | T): Promise<T> {
    const task = this.queue.then(operation, operation);
    this.queue = task.then(() => undefined, () => undefined);
    return task;
  }

  private admitBudget(root: BudgetRoot): StoredBudget {
    reservationStateAdmission(root.state, this.selectedPin);
    const calculatedStateDigest = digest('reservation-state', root.state);
    const calculatedRegistryDigest = digest('reservation-registry', root.state.core.configuration);
    const anchors = [...root.affected_anchors].sort(byteCompare);
    if (root.state_digest !== calculatedStateDigest || root.registry_digest !== calculatedRegistryDigest || !same(anchors, root.affected_anchors) || new Set(anchors).size !== anchors.length) throw new AdmissionError('STATE_INVALID');
    return { ...clone(root), state_bytes: canonical(root.state) };
  }

  async install(request: Row, budgetRoots: BudgetRoot[] = []): Promise<Row> {
    const supplied = clone(request), suppliedBudgets = clone(budgetRoots);
    return this.enqueue(() => {
      const calculated = deploy(supplied, this.selectedPin);
      if (calculated.status !== 'DEPLOYED') return clone(calculated);
      const organization = supplied.deployment_authorization.organization_id as string;
      const key = rootKey(organization, supplied.instance_id);
      const admittedBudgets = suppliedBudgets.map(row => this.admitBudget(row));
      const registryDigests = admittedBudgets.map(row => row.registry_digest).sort(byteCompare);
      const installationBudgetRoots = canonical(admittedBudgets.map(row => ({
        registry_digest: row.registry_digest,
        affected_anchors: clone(row.affected_anchors),
        state: clone(row.state),
        state_digest: row.state_digest,
      })).sort((a, b) => byteCompare(a.registry_digest, b.registry_digest)));
      if (new Set(registryDigests).size !== registryDigests.length) throw new AdmissionError('STATE_INVALID');
      const current = this.roots.get(key);
      if (current) {
        const replay = current.deploy_request === canonical(supplied) && current.state_bytes === canonical(calculated.state) && current.installation_budget_roots === installationBudgetRoots;
        const result = replay ? {
          status: 'HOST_DEPLOY', disposition: 'INSTALLED', organization, instance_id: supplied.instance_id,
          state_digest: current.state_digest, replay: true, dispatch_performed: false,
        } : {
          status: 'HOST_DEPLOY', disposition: 'INSTANCE_OCCUPIED', organization, instance_id: supplied.instance_id,
          authoritative_state_digest: current.state_digest, proposed_state_digest: calculated.state_digest, replay: false, dispatch_performed: false,
        };
        requireShape('lifecycle.schema.json', 'hostDeploymentResult', result);
        return result;
      }
      for (const budget of admittedBudgets) {
        const stored = this.budgets.get(budgetKey(organization, budget.registry_digest));
        if (stored && (stored.state_digest !== budget.state_digest || stored.state_bytes !== budget.state_bytes)) throw new AdmissionError('REVISION_CONFLICT');
      }
      const nextRoot = {
        deploy_request: canonical(supplied), state: clone(calculated.state), state_bytes: canonical(calculated.state), state_digest: calculated.state_digest,
        registry_digests: registryDigests, installation_budget_roots: installationBudgetRoots, event_log: new Map(),
      };
      const result = { status: 'HOST_DEPLOY', disposition: 'INSTALLED', organization, instance_id: supplied.instance_id, state_digest: calculated.state_digest, replay: false, dispatch_performed: false };
      requireShape('lifecycle.schema.json', 'hostDeploymentResult', result);
      for (const budget of admittedBudgets) this.budgets.set(budgetKey(organization, budget.registry_digest), budget);
      this.roots.set(key, nextRoot);
      return result;
    });
  }

  private budgetSnapshot(organization: string, root: StoredRoot): BudgetRoot[] {
    return root.registry_digests.map(registryDigest => {
      const row = this.budgets.get(budgetKey(organization, registryDigest));
      if (!row) throw new Error('Missing authoritative budget root');
      return { registry_digest: row.registry_digest, affected_anchors: clone(row.affected_anchors), state: clone(row.state), state_digest: row.state_digest };
    });
  }

  async snapshot(organization: string, instanceId: string): Promise<{ state: Row; state_digest: string; budget_states: BudgetRoot[] }> {
    return this.enqueue(() => {
      const root = this.roots.get(rootKey(organization, instanceId));
      if (!root) throw new Error('Instance is not installed');
      return { state: clone(root.state), state_digest: root.state_digest, budget_states: this.budgetSnapshot(organization, root) };
    });
  }

  private statefulRefusal(organization: string, root: StoredRoot, request: Row, code: string): Row {
    const result = { status: 'REFUSED', profile: request.profile, role: request.role, code, path: '', state: clone(root.state), state_digest: root.state_digest, budget_states: this.budgetSnapshot(organization, root) };
    requireShape('lifecycle.schema.json', 'statefulRefusal', result);
    return result;
  }

  async submit(organization: string, request: Row, failureReference: string | null = null): Promise<Row> {
    const supplied = clone(request);
    return this.enqueue(() => {
      const key = rootKey(organization, supplied.state.instance_id);
      const root = this.roots.get(key);
      if (!root) throw new Error('Instance is not installed');
      const event = supplied.event as Row, retained = root.event_log.get(event.event_id);
      if (retained && retained.event !== canonical(event)) return this.statefulRefusal(organization, root, supplied, 'REPLAY_CONFLICT');
      if (supplied.state_digest !== root.state_digest || canonical(supplied.state) !== root.state_bytes) return this.statefulRefusal(organization, root, supplied, 'REVISION_CONFLICT');
      if (retained) {
        const replay = step(supplied, this.selectedPin);
        if (replay.status !== 'STEP' || replay.replay !== true) throw new Error('Consumer did not return the retained exact replay');
        return { transition: clone(replay), host_commit: null };
      }
      const budgetInputs = ['PROPOSE', 'EFFECT_OBSERVED'].includes(event.kind) ? event.budget_inputs : [];
      for (const input of budgetInputs as Row[]) {
        const budget = this.budgets.get(budgetKey(organization, input.registry_digest));
        if (!budget || input.request.state_digest !== budget.state_digest || canonical(input.request.state) !== budget.state_bytes) return this.statefulRefusal(organization, root, supplied, 'REVISION_CONFLICT');
      }
      const transition = step(supplied, this.selectedPin);
      if (transition.status !== 'STEP') return clone(transition);
      const before = (budgetInputs as Row[]).map(input => ({ registry_digest: input.registry_digest, affected_anchors: clone(input.affected_anchors), state_digest: input.request.state_digest })).sort((a, b) => byteCompare(a.registry_digest, b.registry_digest));
      if (failureReference !== null) {
        const failed = { status: 'HOST_COMMIT', disposition: 'COMMIT_FAILED', transaction_digest: transition.transaction_digest, state_before_digest: supplied.state_digest, authoritative_state_digest: root.state_digest, budget_before: before, authoritative_budgets: clone(before), failure_reference: failureReference, dispatch_performed: false };
        requireShape('lifecycle.schema.json', 'hostCommitResult', failed);
        return { transition: clone(transition), host_commit: failed };
      }
      const admittedBudgets = (transition.budget_results as Row[]).map(row => this.admitBudget({ registry_digest: row.registry_digest, affected_anchors: clone(row.affected_anchors), state: clone(row.result.state), state_digest: digest('reservation-state', row.result.state) }));
      const after = admittedBudgets.map(row => ({ registry_digest: row.registry_digest, affected_anchors: clone(row.affected_anchors), state_digest: row.state_digest })).sort((a, b) => byteCompare(a.registry_digest, b.registry_digest));
      const registryDigests = [...new Set([...root.registry_digests, ...admittedBudgets.map(row => row.registry_digest)])].sort(byteCompare);
      const eventLog = new Map(root.event_log);
      eventLog.set(event.event_id, { event_digest: digest('runtime-event', event), transaction_digest: transition.transaction_digest, event: canonical(event), request: canonical(supplied), response: canonical(transition) });
      const nextRoot: StoredRoot = { ...root, state: clone(transition.state), state_bytes: canonical(transition.state), state_digest: transition.state_digest, registry_digests: registryDigests, event_log: eventLog };
      const committed = { status: 'HOST_COMMIT', disposition: 'COMMITTED', transaction_digest: transition.transaction_digest, state_before_digest: supplied.state_digest, state_after_digest: transition.state_digest, budget_before: before, budget_after: after, dispatch_performed: false };
      requireShape('lifecycle.schema.json', 'hostCommitResult', committed);
      for (const admitted of admittedBudgets) this.budgets.set(budgetKey(organization, admitted.registry_digest), admitted);
      this.roots.set(key, nextRoot);
      return { transition: clone(transition), host_commit: committed };
    });
  }

  async dispatchPermitted<T>(organization: string, instanceId: string, permitId: string, nativeRequest: Row, connector: (request: Row) => Promise<T> | T): Promise<T> {
    const supplied = clone(nativeRequest);
    return this.enqueue(async () => {
      const root = this.roots.get(rootKey(organization, instanceId));
      if (!root) throw new Error('Instance is not installed');
      const permit = root.state.permits.find((row: Row) => row.permit_id === permitId);
      const active = root.state.active.find((row: Row) => row.occurrence.occurrence_id === permit?.occurrence.occurrence_id);
      const superseded = root.state.permits.some((row: Row) => row.supersedes_permit_digest === permitId);
      const clocks = permit ? root.state.clocks.filter((row: Row) => row.status === 'AVAILABLE' && row.source === permit.clock_source).sort((a: Row, b: Row) => BigInt(a.revision) < BigInt(b.revision) ? -1 : 1) : [];
      const now = clocks.at(-1)?.observed_time;
      if (!permit || !active || active.status !== 'PERMITTED' || superseded || !same(permit.native_request, supplied) || !now || instant(now) >= instant(permit.expires_at)) throw new AdmissionError('BINDING_MISMATCH');
      const claim = canonical([organization, instanceId, permitId]);
      if (this.dispatchClaims.has(claim)) throw new AdmissionError('PREREQUISITE_MISSING');
      this.dispatchClaims.add(claim);
      return connector(clone(supplied));
    });
  }
}
